src/examples/zkapps/voting/membership.ts

```diff
@@ -6,31 +6,175 @@ import {
   method,
   DeployArgs,
   Permissions,
+  Bool,
+  PublicKey,
+  Experimental,
+  Circuit,
 } from 'snarkyjs';
+import { Member } from './member.js';
+import { ParticipantPreconditions } from './preconditions.js';
 
-export class Membership extends SmartContract {
-  // TODO: Add state variables
+let participantPreconditions = ParticipantPreconditions.default;
+
+interface MembershipParams {
+  participantPreconditions: ParticipantPreconditions;
+  contractAddress: PublicKey;
+  doProofs: boolean;
+}
+
+/**
+ * Returns a new contract instance that based on a set of preconditions.
+ * @param params {@link MembershipParams}
+ */
+export async function Membership(
+  params: MembershipParams
+): Promise<Membership_> {
+  participantPreconditions = params.participantPreconditions;
+
+  let contract = new Membership_(params.contractAddress);
+  if (params.doProofs) {
+    await Membership_.compile();
+  }
+
+  return contract;
+}
+
+/**
+ * The Membership contract keeps track of a set of members.
+ * The contract can either be of type Voter or Candidate.
+ */
+export class Membership_ extends SmartContract {
+  /**
+   * Root of the merkle tree that stores all committed members.
+   */
+  @state(Field) committedMembers = State<Field>();
+
+  /**
+   * Accumulator of all emitted members.
+   */
+  @state(Field) accumulatedMembers = State<Field>();
+
+  reducer = Experimental.Reducer({ actionType: Member });
 
   deploy(args: DeployArgs) {
     super.deploy(args);
     this.setPermissions({
       ...Permissions.default(),
-      editState: Permissions.proofOrSignature(),
+      editState: Permissions.none(), // TODO: fix permissions
+      editSequenceState: Permissions.none(), // TODO: fix permissions
     });
-    // TODO: Add account state initilaztion here
   }
 
-  @method addEntry(accountId) {
+  /**
+   * Method used to add a new member.
+   * Dispatches a new member sequence event.
+   * @param member
+   */
+  @method addEntry(member: Member): Bool {
     // Emit event that indicates adding this item
     // Preconditions: Restrict who can vote or who can be a candidate
+
+    // since we need to keep this contract "generic", we always assert within a range
+    // even tho voters cant have a maximum balance, only candidates
+    // but for a voter we simply use UInt64.MAXINT() as the maximum
+    let party = Experimental.createChildAccountUpdate(
+      this.self,
+      member.publicKey
+    );
+    party.account.balance.assertEquals(party.account.balance.get());
+
+    let balance = party.account.balance.get();
```

Comment:
can you grep for "party" in these new file an rename to accountUpdate?

Reply:
ahhh yeep, good catch!